plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+	GCancellable *cancellable = gs_app_get_cancellable (get_os_upgrade (plugin));
+
+	if (priv->os_upgrade_cancellable == cancellable)
+		return;
+
+	g_clear_object (&priv->os_upgrade_cancellable);
+	priv->os_upgrade_cancellable = g_object_ref (cancellable);
+	g_cancellable_connect (priv->os_upgrade_cancellable,
+			       G_CALLBACK (os_upgrade_cancelled_cb),
+			       plugin, NULL);
+}
+
+static void
+os_upgrade_set_download_by_user (GsApp *app, gboolean value)
+{
+	g_autoptr(GVariant) var = g_variant_new_boolean (value);
+	/* we need to assign it to NULL in order to be able to override it
+	 * (safeguard mechanism in GsApp...) */
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", NULL);
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", var);
+}
+
+static gboolean
+os_upgrade_get_download_by_user (GsApp *app)
+{
+	GVariant *value = gs_app_get_metadata_variant (app, "eos::DownloadByUser");
+	if (value == NULL)
+		return FALSE;
+	return g_variant_get_boolean (value);
+}
+
+static void
+os_upgrade_force_fetch (EosUpdater *updater)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	eos_updater_call_fetch_full (updater,
+				     g_variant_dict_end (&options_dict),
+				     NULL, NULL, NULL);
+}
+
+static gboolean
+os_upgrade_force_fetch_sync (EosUpdater *updater,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	return eos_updater_call_fetch_full_sync (updater,
+						 g_variant_dict_end (&options_dict),
+						 cancellable, error);
+}
+
+/* This method deals with the synchronization between the EOS updater's states
+ * (DBus service) and the OS upgrade's states (GsApp), in order to show the user
+ * what is happening and what they can do. */
+static EosUpdaterState
+sync_state_from_updater (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *app = get_os_upgrade (plugin);
+	EosUpdaterState state;
+	AsAppState previous_app_state = gs_app_get_state (app);
+	AsAppState current_app_state;
+	const gchar *version;
+
+	state = eos_updater_get_state (priv->updater_proxy);
+	g_debug ("EOS Updater state changed: %s", eos_updater_state_str [state]);
+
+	switch (state) {
+	case EOS_UPDATER_STATE_NONE:
+	case EOS_UPDATER_STATE_READY:
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+		break;
+	case EOS_UPDATER_STATE_POLLING:
+		break;
+	case EOS_UPDATER_STATE_UPDATE_AVAILABLE: {
+		guint64 total_size = 0;
+		if (gs_app_get_state (app) == AS_APP_STATE_INSTALLING) {
+			/* when the OS upgrade is already being installed and the
+			 * updater reports an available update, (meaning we were
+			 * polling before), we should readily call fetch */
+			os_upgrade_force_fetch (priv->updater_proxy);
+			break;
+		}
+
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
```

Comment:
I would prefer if this code was in an `else` block and you had a single `break` at the end of the `case`

Reply:
I did it like this because we use early returns everywhere, and it's the same principle. But I can change it...